### PR TITLE
Allow to change the cluster default network type

### PR DESCRIPTION
**Pull request by pliurh**
Allow users to change the cluster default network type by updating the 'networkType' field of the network.config.openshift.io CR. This not only support the use case of migrating the cluster network from Openshift SDN to OVN-Kubernetes, but also the use case vise Versa.

### CODE REVIEW

**Comment by danwinship** [comment]
/hold
We can't allow this in general; it would completely break the cluster. We need to have some way of allowing it as part of the supported migration procedure, but blocking it otherwise.

File: bindata/network/openshift-sdn/sdn-ovs.yaml [hunk 1]
```diff
@@ -132,7 +132,14 @@ spec:
...
             exec:
-              command: ["/usr/share/openvswitch/scripts/ovs-ctl", "stop"]
+              command: 
+              - /bin/bash
+              - -c
+              - |
+                #!/bin/bash
+                /usr/share/openvswitch/scripts/ovs-ctl stop
+                rm -rf /run/openvswitch/*
+                rm -rf /var/lib/openvswitch/*
```
Comment (danwinship) [inline bindata/network/openshift-sdn/sdn-ovs.yaml]: We don't want to do this on ordinary termination; it would mean that every upgrade would tear down the network. (Likewise for the ovn-kube change.)

File: bindata/network/openshift-sdn/sdn.yaml [hunk 2]
```diff
@@ -55,6 +55,10 @@ spec:
       containers:
       - name: sdn
         image: {{.SDNImage}}
+        lifecycle:
+          preStop:
+            exec:
+              command: ["rm","-f","/etc/cni/net.d/80-openshift-network.conf"]
```
Comment (danwinship) [inline bindata/network/openshift-sdn/sdn.yaml]: there is already a `preStop` handler at the end of the container spec that does this

File: bindata/network/ovn-kubernetes/monitor.yaml [hunk 3]
```diff
@@ -76,9 +76,9 @@ spec:
   publishNotReadyAddresses: true
   ports:
   - name: metrics
-    port: 9101
+    port: 9103
```
Comment (danwinship) [inline bindata/network/ovn-kubernetes/monitor.yaml]: (this could be done as a separate PR...)

File: pkg/controller/statusmanager/status_manager.go [hunk 4]
```diff
@@ -90,6 +90,99 @@ func (status *StatusManager) deleteRelatedObjectsNotRendered(co *configv1.Cluste
 			objToDelete.SetName(currentObj.Name)
 			objToDelete.SetNamespace(currentObj.Namespace)
 			objToDelete.SetGroupVersionKind(gvk)
+			if objToDelete.GetKind() == "Namespace" {
+				continue
```
Comment (danwinship) [inline pkg/controller/statusmanager/status_manager.go]: Why was this added? We want to delete unused namespaces...
(Though... I guess probably we need to fix it to do that *last*.)

File: pkg/controller/statusmanager/status_manager.go [hunk 5]
```diff
@@ -114,6 +207,8 @@ func (status *StatusManager) set(reachedAvailableLevel bool, conditions ...confi
 
 		oldStatus := co.Status.DeepCopy()
 		status.deleteRelatedObjectsNotRendered(co)
+		status.deleteDaemonSetNotRendered(co)
+		status.deleteDeploymentNotRendered(co)
```
Comment (danwinship) [inline pkg/controller/statusmanager/status_manager.go]: why is this needed? The DaemonSets and Deployments should also be in RelatedObjects

**Comment by pliurh** [comment]
> /hold
> We can't allow this in general; it would completely break the cluster. We need to have some way of allowing it as part of the supported migration procedure, but blocking it otherwise.

Agree. What's your suggestion on how to do it? How about adding an annotation like `maintenance: "migration"` in the `cluster` CR to indicate if this kind of operation can be allowed or not?

File: bindata/network/ovn-kubernetes/monitor.yaml (hunk 3 above)
Comment (pliurh) [reply bindata/network/ovn-kubernetes/monitor.yaml]: Will do.

File: pkg/controller/statusmanager/status_manager.go (hunk 4 above)
Comment (pliurh) [reply pkg/controller/statusmanager/status_manager.go]: Due to some reason, the 'namespace' deletion is always stuck at the "Terminating" state forever, even when all the related objects in that namespace have been deleted. And it will block the same Namespace creation if we what to perform a rollback from ovn-kubernetes to openshift-sdn. 

I agree that we need to remove the unused namespaces anyway. But I think maybe we can fix it later.

Comment (danwinship) [reply pkg/controller/statusmanager/status_manager.go]: > Due to some reason, the 'namespace' deletion is always stuck at the "Terminating" state forever, even when all the related objects in that namespace have been deleted.

Maybe there is something else in that Namespace that isn't in the related objects then?

**Comment by danwinship** [comment]
An annotation sounds good. It should be more namespaced than that though. Maybe something like

    "networkoperator.openshift.io/network-migration": ""

(Standard annotation semantics are that if there's nothing to say about the annotation beyond the fact that it exists, then you just give it an empty-string value rather than `"true"` or something like that.)

You still need to figure out exactly what that entails though; you don't want CNO to start changing DaemonSets right away because that would interfere with the controlled node-by-node migration plan... do you stop the CNO during the migration? Is the rest of the code for dealing with migrations posted somewhere?

File: bindata/network/openshift-sdn/sdn-ovs.yaml (hunk 1 above)
Comment (pliurh) [reply bindata/network/openshift-sdn/sdn-ovs.yaml]: fixed.

File: bindata/network/openshift-sdn/sdn.yaml (hunk 2 above)
Comment (pliurh) [reply bindata/network/openshift-sdn/sdn.yaml]: fixed.

File: pkg/controller/statusmanager/status_manager.go (hunk 5 above)
Comment (pliurh) [reply pkg/controller/statusmanager/status_manager.go]: The origin intention is to remove the deployment and daemonset before other objects. But I found it is unnecessary after more tests. It shall be removed.

**Comment by pliurh** [comment]
> An annotation sounds good. It should be more namespaced than that though. Maybe something like
> 
> ```
> "networkoperator.openshift.io/network-migration": ""
> ```
> 
> (Standard annotation semantics are that if there's nothing to say about the annotation beyond the fact that it exists, then you just give it an empty-string value rather than `"true"` or something like that.)
> 
> You still need to figure out exactly what that entails though; you don't want CNO to start changing DaemonSets right away because that would interfere with the controlled node-by-node migration plan... do you stop the CNO during the migration? Is the rest of the code for dealing with migrations posted somewhere?

In 4.4, we plan to take the cluster reboot approach, which means all the nodes will be rebooted during the migration. Therefore, the whole cluster shall be considered as out of service during the migration. So all the pod will be connected to the new network backend after reboot. I gave a live demo at the Group C demo on Jan 8th, however, the demo record has not been added to the agenda doc yet. But here is a previous [demo record](https://drive.google.com/file/d/1AuyTzuJ0xjI3vhyvOQD72vwPtDxp2zUT/view), which shows the whole procedure of the 'cluster reboot' migration.

File: pkg/controller/operconfig/operconfig_controller.go [hunk 6]
```diff
@@ -31,6 +31,11 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/source"
 )
 
+const (
+	// The Network Migration Annotation which indicates if the operator is allowed to change the cluster network type.
+	MigrationAnno = "networkoperator.openshift.io/network-migration"
```
[the remark this answers is not in the document]
Comment (pliurh) [reply pkg/controller/operconfig/operconfig_controller.go]: fixed.

File: pkg/controller/statusmanager/status_manager.go (hunk 4 above)
Comment (pliurh) [reply pkg/controller/statusmanager/status_manager.go]: Yes, there are some objects which are not managed by the CNO. I am not sure what is the proper way to delete all of them, then delete the namespace itself.

Comment (danwinship) [reply pkg/controller/statusmanager/status_manager.go]: hm. do you remember what objects?

Comment (pliurh) [reply pkg/controller/statusmanager/status_manager.go]: I did some further tests. It turns out that the namespace will be stuck in 'Terminating' before the cluster reboot, but can be deleted eventually after that. Therefore I remove the lines above.

**Comment by danwinship** [comment]
Thanks. Can you squash the commits?

**Comment by pliurh** [comment]
> Thanks. Can you squash the commits?

Done.

**Comment by pliurh** [comment]
> So this is looking pretty simple now. We still need to figure out what's going wrong with deleting the namespace

Before rebooting, the cluster is in an unstable status. The stuck namespace has a finalizer 'kubernetes', which blocks the deletion. I think it should be fine, as long as, the namespace can be deleted successfully after the cluster back to normal.
